### PR TITLE
Encoder Wrapper Improvements

### PR DESCRIPTION
**Pull request by DeNeutoy**
These wrappers need some better tests if they're going to stay like this - thoughts?

### CODE REVIEW

File: allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        self._module = module
+
+    def get_output_dim(self) -> int:
+        return self._module.hidden_size * (2 if self._module.bidirectional else 1)
+
+    def forward(self,  # pylint: disable=arguments-differ
+                inputs: torch.Tensor,
+                sequence_lengths: torch.LongTensor = None) -> torch.Tensor:
+
+        if sequence_lengths is None:
```
Comment (matt-gardner) [inline allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py]: What's the motivation for having this check?  Won't the call below fail if it's given a tensor, instead of a packed sequence?

File: allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py [hunk 2]
```diff
@@ -19,11 +21,27 @@ class WrappedPytorchRnn(Seq2VecEncoder):
...
+                sequence_lengths: torch.LongTensor = None) -> torch.Tensor:
+
+        if sequence_lengths is None:
+            return self._module(inputs)[0][:, :, -1]
+        sorted_inputs, sorted_sequence_lengths, restoration_indices = sort_batch_by_length(inputs,
+                                                                                           sequence_lengths)
+        packed_sequence_input = pack_padded_sequence(sorted_inputs, sorted_sequence_lengths, batch_first=True)
+
+        # Actually call the module on the sorted PackedSequence.
+        _, state = self._module(packed_sequence_input)
```
Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: This does not have the same behavior as the case above.  The `state` returned as the second argument is `(batch_size, num_layers * num_directions, hidden_size)`, which is not actually what we want.

File: allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py (hunk 1 above)
Comment (DeNeutoy) [inline allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py]: No - you can use both: if you don't have padded sequences, you can just call the rnn on the tensor directly.

File: allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py (hunk 2 above)
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: Ok - what do you want this to return?

Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: Which one do you think is better?  We definitely want the return shape to be `(batch_size, output_dim)`, however we get there.

Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: I guess for now it's easiest to just take the last vector of the last layer?

Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: I think longer term we want this to be configurable (e.g., do you concat the two directions, sum them, max pool them, max pool the state vectors for each timestep, etc.), but for now, I'd say just pick one that seems reasonable and make sure it's consistent in both cases.

File: allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py (hunk 1 above)
Comment (matt-gardner) [inline allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py]: So if you're doing language modeling, and you've set it up that all of your sequences are already the same length?  Is that the case you're talking about?

File: allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    def get_output_dim(self) -> int:
+        return self._module.hidden_size * (2 if self._module.bidirectional else 1)
+
+    def forward(self,  # pylint: disable=arguments-differ
+                inputs: torch.Tensor,
+                sequence_lengths: torch.LongTensor = None) -> torch.Tensor:
+
+        if sequence_lengths is None:
+            return self._module(inputs)[0][:, -1, :]
```
Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: I think you're missing a transpose here, so this is returning the state for the last item in the batch.  I _think_ this should be `return self._module(inputs)[0][-1, :, :]`.

File: allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            state = state[0]
+
+        # Restore the original indices and return the final state of the
+        # top layer. Pytorch's recurrent layers return state in the form
+        # (num_layers * num_directions, batch_size, hidden_size) regardless
+        # of the 'batch_first' flag, so we transpose, extract the relevant
+        # layer state (both forward and backward if using bidirectional layers)
+        # and return them as a single (batch_size, self.get_output_dim()) tensor.
+
+        # Pytorch doesn't respect batch first return states, so we transpose
```
Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: You're repeating yourself a little bit in these two comment blocks.

File: allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        # here to keep things batch first. unsorted_state is now of shape:
+        # (batch_size, num_layers * num_directions, hidden_size).
+        unsorted_state = state.transpose(0, 1)[restoration_indices]
+        last_state_index = 2 if self._module.bidirectional else 1
+
+        last_layer_state = unsorted_state[:, -last_state_index:, :]
```
Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: I was about to say that this was going to give you the wrong shape output, but you do a reshape.  Then I was confused about how reshaping something of `(batch, 2, hidden_size)` would work when reshaped to `(batch, output_dim)`.  I think there should be a comment in the class docstring that we concatenate the forward and backward states, and a reminder / explanatory comment here that says what we're doing.

File: tests/modules/seq2vec_encoders/pytorch_seq2vec_wrapper_test.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        tensor[3, 2:, :] = 0
+        tensor[4, 1:, :] = 0
+
+        input_tensor = Variable(tensor)
+        sequence_lengths = torch.LongTensor([7, 6, 4, 2, 1])
+        packed_sequence = pack_padded_sequence(input_tensor, list(sequence_lengths), batch_first=True)
+        _, state = lstm(packed_sequence)
+        # Transpose output state, extract the last forward and backward states and
+        # reshape to be of dimension (batch_size, 2 * hidden_size).
+        reshaped_state = state[0].transpose(0, 1)[:, -2:, :].contiguous().view(-1, 14)
```
Comment (matt-gardner) [inline tests/modules/seq2vec_encoders/pytorch_seq2vec_wrapper_test.py]: Can we test against an explicit concatenation, instead of a reshaping?  It'd be nice to be absolutely sure that the reshaping is happening as we expect it is.

File: allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py (hunk 1 above)
Comment (DeNeutoy) [inline allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py]: yes, or if you're doing something which doesn't require batching, like running an LSTM over a fixed size number of memory slots or something like they did in that Dynamic Memory Networks paper. I can remove this if you think it's confusing.

File: allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py (hunk 5 above)
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: Agreed.

File: allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py (hunk 3 above)
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: Don't think so - see http://pytorch.org/docs/0.1.12/nn.html#torch.nn.RNN. The `batch_first` flag effects the input and output _sequences_, but not the hidden states. Although it depends if we want to support non - batch_first RNNs. I would say no in general, as our data api doesn't support this at all.

Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: But you do a transpose below, right?

Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: Yes, because the state returned from the RNN is not batch_first, regardless of whether the RNN class was instantiated with the `batch_first` flag.

Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: Oh, I was confused here.  You're doing `self._module(inputs)[0]`, not `self._module(inputs)[1]`.  Are these guaranteed to return the same thing?  Seems like it'd be easier to understand the code if both branches had similar logic.

Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: Or at least a nice comment explaining the decision.

Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: Yeah sorry, this could have been a lot clearer - this is actually important, because they are _not_ the same _if_ you have variable length sequences. By using the state, pytorch collects the correct final state for each sequence, but if you just take the last element from `lstm(inputs)[0]`, this will be zeros for elements which are not as long as the batch. However, if you aren't passing sequence lengths in, we're assuming that your sequences are full, in which case they are the same (and it's clearly a simpler operation, which was why I preferred it here).

Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py]: Also, I added this switch because apparently packing and padding can be a bottleneck, looks like something Pytorch are trying to improve : https://github.com/pytorch/pytorch/issues/1788
